Add grindstone event, use ItemStack-sensitive getMaxDamage

SilentChaos512 opened the pull request:
This makes a change to `GrindstoneContainer` so it uses `ItemStack.getMaxDamage` instead of `Item.getMaxDamage`, which makes the results more accurate for some items. I also added an event for mods which need more control over the output. The event is very similar to `AnvilUpdateEvent`. It can also be canceled, which will clear the grindstone's output slot.

**A simple example mod can be found here**: https://gist.github.com/SilentChaos512/74a31a24e348fbe5c752b184fa394d9c  
This shows both overriding the output and canceling the event.

SilentChaos512 committed:
Add GrindstoneUpdateEvent, use ItemStack.getMaxDamage in grindstone

LexManos commented:
This isn't bad, but the worry is the experience values and the fact that they are calculated in the pull. And could not be controlled by this event.
Perhaps it would be worth redirecting the exp value to a field on the result slot, that is set in this calculate event. And then queried in the consume call. 
Also, would it be feasible to provide the default calculated output as a input to the event? So that modders do not have to re-implement that logic to change a property of the output.

Side note: The change to getMaxDamage is unquestioned, but sadly you decided to combine these two completely separate changes into one PR so I can't accept that without accepting the event which doesn't look like it's anywhere near being done.

SilentChaos512 committed:
Split grindstone event into two, add XP override

Pre event occurs where the old event did. Post event occurs much later, and provides the vanilla output.

SilentChaos512 commented:
Added a method to override the XP produced by the grindstone. The vanilla XP value is not provided to the event, is that acceptable?

I also split the event into two, one which is fired early on to bypass most vanilla logic, and a second which provides the vanilla output for simple changes.

Small issue with the previous patch: `int l = itemstack.func_77958_k() - itemstack1.func_77952_i();` needs to be `int l = itemstack1.func_77958_k() - itemstack1.func_77952_i();` instead. Can this be edited to resolve the conflict?

LexManos commented on patches/minecraft/net/minecraft/inventory/container/GrindstoneContainer.java.patch:
```diff
@@ -0,0 +1,58 @@
+    };
+    private final IWorldPosCallable field_217015_e;
++   public int xpOnCraft = -1;
+ 
```
Ha... No.... public field is horrible.
Best would probably be to have the same system Furnaces do, where the crafting action adds to a internal xp pool, that is dumped when there is a item pulled.

LexManos commented on patches/minecraft/net/minecraft/inventory/container/GrindstoneContainer.java.patch:
```diff
@@ -0,0 +1,58 @@
+          }
+ 
+-         this.field_217013_c.func_70299_a(0, this.func_217007_a(itemstack2, i, j));
++         net.minecraftforge.common.ForgeHooks.onGrindstoneChangePost(itemstack, itemstack1, this.func_217007_a(itemstack2, i, j), field_217013_c, this);
```
The grindstone changing shouldn't mutate the input stacks... so cant you just fire one event at the end? Ya it'd be some code running when unneeded, but you'd have the vanilla output to work with. And you wouldn't be firing two events.
